libp2p/autotls/peeridauth.nim

```diff
@@ -0,0 +1,202 @@
+import base64, json, random, strutils
+import chronos/apps/http/httpclient, results, chronicles, bio
+import ./utils, ../peerinfo, ../crypto/crypto
+
+logScope:
+  topics = "libp2p peerid auth"
+
+type SigParam = object
+  k: string
+  v: seq[byte]
+
+proc randomChallenge(): string =
+  randomize()
+  let charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+  result = ""
+  for _ in 0 ..< 48:
+    result.add sample(charset)
+
+proc extractField(data, key: string): string {.raises: [PeerIDAuthError].} =
+  # Helper to extract quoted value from key
+  for segment in data.split(","):
+    if key in segment:
+      return segment.split("=", 1)[1].strip(chars = {' ', '"'})
+  raise newException(PeerIDAuthError, "Could not find " & key & " in " & data)
+
+proc encodeVarint(n: int): seq[byte] =
+  var varInt: seq[byte] = @[]
+  var x = uint64(n)
+  while x != 0:
+    var byteVal = byte(x and 0x7F)
+    x = x shr 7
+    if x != 0:
+      byteVal = byteVal or 0x80
+    varInt.add(byteVal)
+  return varInt
+
+proc genDataToSign(prefix: string, parts: seq[SigParam]): seq[byte] =
+  var buf: seq[byte] = prefix.toByteSeq()
+  for p in parts:
+    buf.add encodeVarint(p.k.len + p.v.len + 1)
+    buf.add (p.k & "=").toByteSeq
+    buf.add p.v
+  return buf
+
+proc peerIdSign(
+    privateKey: PrivateKey,
+    challenge: string,
+    publicKey: PublicKey,
+    hostname: string,
+    clientSender: bool = true,
+): string =
+  let prefix = "libp2p-PeerID"
+  let parts =
+    # parts need to be sorted alphabetically by key
+    if clientSender:
+      @[
+        SigParam(k: "challenge-client", v: challenge.toByteSeq()),
+        SigParam(k: "hostname", v: hostname.toByteSeq()),
+        SigParam(k: "server-public-key", v: publicKey.getBytes().get()),
+      ]
+    else:
+      @[
+        SigParam(k: "challenge-server", v: challenge.toByteSeq()),
+        SigParam(k: "client-public-key", v: publicKey.getBytes().get()),
+        SigParam(k: "hostname", v: hostname.toByteSeq),
+      ]
+  let bytesToSign = genDataToSign(prefix, parts)
+  return base64.encode(privateKey.sign(bytesToSign).get().getBytes(), safe = true)
+
+proc checkSignature(
+    serverSig: string,
+    serverPublicKey: PublicKey,
+    challengeServer: string,
+    clientPublicKey: PublicKey,
+    hostname: string,
+): bool {.raises: [PeerIDAuthError].} =
+  let prefix = "libp2p-PeerID"
+  let parts: seq[SigParam] =
+    @[
+      SigParam(k: "challenge-server", v: challengeServer.toByteSeq()),
+      SigParam(k: "client-public-key", v: clientPublicKey.getBytes().get()),
+      SigParam(k: "hostname", v: hostname.toByteSeq),
+    ]
+  let signedBytes = genDataToSign(prefix, parts)
```

Comment:
This is very similar to code prev function. Could we extract this into a separate function?